Source/Hypermedia.Client/HypermediaClientBuilder.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using Bluehands.Hypermedia.Client.Hypermedia;
+using Bluehands.Hypermedia.Client.ParameterSerializer;
+using Bluehands.Hypermedia.Client.Reader;
+using Bluehands.Hypermedia.Client.Resolver;
+
+namespace Bluehands.Hypermedia.Client
+{
+    public class HypermediaClientBuilder
+    {
+        private Func<IHypermediaObjectRegister> createHypermediaObjectRegister;
+        private Func<IParameterSerializer> createParameterSerializer;
+        private Func<IStringParser> createStringParser;
+        private Func<IProblemStringReader> createProblemStringReader;
+        private Func<IParameterSerializer, IProblemStringReader, IHypermediaResolver> createHypermediaResolver;
+        private Func<IHypermediaObjectRegister, IStringParser, IHypermediaReader> createHypermediaReader;
+
+        public HypermediaClientBuilder()
+        {
+
+        }
+
+        private static T Get<T>(Func<T> getFunc, params string[] possibleMethodNames)
+        {
+            if (getFunc == null)
+            {
+                throw new InvalidOperationException(
+                    $"Please call any of {string.Join(",", possibleMethodNames)}, or a suitable extension method from one of the Hypermedia.Client.Extensions packages, before creating the HypermediaClient");
+            }
+
+            return getFunc();
+        }
+
+        public HypermediaClient<TEntryPoint> CreateHypermediaClient<TEntryPoint>(Uri uriApiEntryPoint) where TEntryPoint : HypermediaClientObject
+        {
+            var objectRegister = Get(this.createHypermediaObjectRegister, nameof(ConfigureObjectRegister));
+            var serializer = Get(this.createParameterSerializer, nameof(WithCustomParameterSerializer));
+            var stringParser = Get(this.createStringParser, nameof(WithCustomStringParser));
+            var problemReader = Get(this.createProblemStringReader, nameof(WithCustomProblemStringReader));
+            var resolver = Get(() => this.createHypermediaResolver(serializer, problemReader), nameof(WithCustomHypermediaResolver));
+            var reader = Get(() => this.createHypermediaReader(objectRegister, stringParser), nameof(WithSirenHypermediaReader), nameof(WithCustomHypermediaReader));
+            resolver.InitializeHypermediaReader(reader);
+            reader.InitializeHypermediaResolver(resolver);
+            var client = new HypermediaClient<TEntryPoint>(uriApiEntryPoint, resolver, reader);
+            return client;
+        }
+
+        public HypermediaClientBuilder ConfigureObjectRegister(Action<HypermediaObjectRegister> configure)
```

Comment:
```suggestion
        public HypermediaClientBuilder ConfigureObjectRegister(Action<IHypermediaObjectRegister> configure)
```
Pass interface of IHypermediaObject register into the action rather than exposing the HypermediaObjectRegister type